README.md
# AutoCirc


## What is AutoCirc
AutoCirc is a computational program to automatically identify back-splice junctions of potential circular RNAs (circRNAs) from RNA-seq data. Its core algorithm is implemented by C++, and it is wrapped in Perl, so it runs faster than most other circRNA detection tools under the same computer environment. AutoCirc is designed to detect circRNAs for all species as long as their genomes are available. It does not require gene annotations, but gene annotations will improve detection of circRNAs. AutoCirc runs on Linux/Unix OS.

Version: 1.3

Last Modified: 08/21/2018

Authors: Chan Zhou (zhou.chan@mgh.harvard.edu)

Maintainer: Chan Zhou 

## A Workflow of AutoCirc 
![pipeline](https://github.com/chanzhou/AutoCirc/blob/master/AutoCirc_Workflow.jpg)

AutoCirc detects circRNAs from the BAM file of unmapped reads (output of Bowtie2 / Bowtie / TopHat) by the following steps: 
1)	Extract the 20 nucleotide anchors from the left and right ends of unmapped reads and map them to the reference genome with Bowtie2. 
2)	Scan the mapped anchors from both ends of unmapped reads to identify forward splice junctions (splicing from the 3’ end of an upstream sequence to the 5’ end of a downstream sequence) and back splice junctions (splicing from the 3’ end of a downstream sequence to the 5’ end of an upstream sequence) with unambiguous GT-AG breakpoint detection. We required a genomic distance between the two splice sites of no more than 100 kb and at least 100 bp (as default values). 
3)	Identify the back splice junctions that meet the annotated exon boundaries (if known gene annotation is provided).
4)	Merge all the back splice junctions either found based on the GT-AG rule of splice donor and acceptor or based on the exon boundaries of annotated genes. 
5)	Remove all forward splice junctions and filter out the pseudo-back splice junctions, which span the 3’ end and 5’ end of the duplicated genes.

Finally, AutoCirc will output the back splice junctions that contain the canonical GT-AG splice donor and acceptor sequences in the flanking intron boundaries or that occur at known exon boundaries (if inputting reference annotation, e.g. RefSeq used here). These back splice junctions are considered to represent circRNAs.

## Prerequisites

To use AutoCirc, you will need the following programs in your PATH:
* Bowtie2 (version 2.2.5 or higher)
* bedtools
* samtools

You will also need Perl version 5.16.3 or higher.


## RNA-seq data
PolyA-depleted, ribosome-depleted, or RNase R-treated RNA-seq data in FASTQ format.
The lengths of reads should be at least 50 nucleotides (nt). 
To identify reliable back splice junctions of potential circRNAs, we recommend using RNA-seq with read lengths of at least 80 nt . 

## Data Preparation
To obtain the unmapped reads, the user should map all reads to a reference genome either by *Bowtie2 or TopHat*. For paired end RNA-seq data, each end of RNA-seq data will be processed separately as a single end RNA-seq data.

For example:
* Bowtie2:

*Step 1: Mapping reads to reference genome*

```bash
bowtie2 -p4 --very-sensitive --mm -D20 --score-min=C,-15,0 -q -x <bowtie2_genome_index> -U <reads.fastq> 1>bowtie2out.sam 2>bowtie2.log 
```

*Step 2: extract the unmapped reads in BAM format*

```bash
samtools view -hbuS bowtie2out.sam | samtools view -hf 4 - | samtools view -Sb - | samtools sort - unmapped
```

### OR

* TopHat:
```bash
tophat -p 4  <options> -o tophat_out <bowtie2_genome_index> <reads.fastq>
```

You will use the *unmapped.bam* file obtained by Bowtie2 or TopHat (in the output folder of TopHat).

## Installation 

### Download AutoCirc
```bash
git clone https://github.com/chanzhou/AutoCirc.git 
cd AutoCirc
chmod 755 AutoCirc_v1.3.pl
chmod 755 script/*
```
Then run the pipeline inside the AutoCirc folder, which includes all the required scripts. 

### Usage
```bash
./AutoCirc_v1.3.pl  [options]

Options:	
	-g/--genome     Absolute path of the reference genome file (Fasta format) 
	-I/--gindex     Absolute path of the genome index of bowtie2 (e.g. /bowtie2_index/hg19)
	--bam		Bam file of unmapped reads (output by bowtie or bowtie2 or tophat)
	-b/--bed	Reference gene annotation (standard bed format) 
			default: without reference gene annotation
			The absolute path to the annotated gene file
	-s/--seed	The size of the seed (e.g. anchor size)
			default: 20 (recommend values: 15-25; this value should be smaller than the (length of reads-10)/2)
	--mis		Maximum allowed mismatches during the search process (0, 1, or 2) 
			default: 0
	--min		Minimum distance between the two splice sites
			default: 100
	--max 		Maximum distance between the two splice sites
			default: 100000
	-o 		Output folder name 
			default: "AutoCirc_output"
	
	-h/--help	This usage
	-v/--version	Print version
```

### Examples :
```bash
./AutoCirc_v1.3.pl  -g /human_genome/hg19.fa -I /bowtie2_index/hg19 --bam unmapped.bam -b /hg19/Annotation/refGene.bed --mis 0 --min 100 --max 100000 -s 20 -o autocirc_output 
```
or

```bash
./AutoCirc_v1.3.pl  -g /human_genome/hg19.fa -I /bowtie2_index/hg19 --bam unmapped.bam -b /hg19/Annotation/refGene.bed
```
or

```bash
./AutoCirc_v1.3.pl  -g /human_genome/hg19.fa --bam unmapped.bam -s 15 -o autocirc_output   
```

## NOTE: 
1. The reference genome file should be the same one used in the previous Botwie/Bowtie2/Tophat mapping.
2. The reference gene annotation should be in standard bed format with full 12 fields ( https://genome.ucsc.edu/FAQ/FAQformat.html#format1) and can be obtained from the UCSC genome browser.
3. The minimum and maximum distance between the two splice sites are suggested to be set as the numbers close to the minimum and the maximum size of annotated genes in the studied genome.  The current default option“--min 100 --max 100000” is set for human genome. If you will study other species, pls change these settings accordingly. 


## Output
AutoCirc will output both the final predicted back splice junction file (circ.final.bed) and also some intermediate tmp files and log files. All are included in the specified output folder (default is “AutoCirc_output”).

The Final output file “circ.final.bed” follows the standard BED format as follows:

| Field(column)| Description                                   |
| :-----------:| :---------------------------------------------|
| Chr	       | Chromosome                                    |
| Start	       | Start of junction                             |
| End	       | End of junction                               |
| Junc_name    | The name of back splice junction              |
| Coverage     | Number of reads mapping to the junction       |
| Site_strand  | Strand of junction                            |
| Distance     | Distance between the two ends of junction     |
| Read_tags    | All the tags of reads mapping to the junction |


## Citation

* Chan Zhou*, Benoit Molinie*, Kaveh Daneshvar, Joshua V. Pondick, Jinkai Wang, Nicholas O. Van Wittenberghe, Yi Xing, Cosmas C. Giallourakis+, and Alan C. Mullen+. Genome-Wide Maps of m6A circRNAs Identify Widespread and Cell-Type-Specific Methylation Patterns that Are Distinct from mRNAs. Cell Reports, 20(9), 2262–2276. doi:10.1016/j.celrep.2017.08.027 (published on Aug 29th, 2017) 

* bioRxiv (March, 2017) version:  http://biorxiv.org/content/early/2017/03/10/115899

## License
Copyright (C) 2017 The Mullen Lab. See the LICENSE file for license rights and limitations.




Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


### [1.3] - 2018-8-21
### Added
- Incorporate the function of removing false positive circRNAs that come across multiple un-overlapped transcripts, but do not originate from another transcripts;
- Include "circ_across_mgenes.pl" in the script folder.

### Changed
- Update the READ file regarding the usage of "--min" and "--max" arguments.

### Removed
-Remove "circ_across2genes.pl" and "circ2mboundaries.pl" from the script folder and related lines in the AutoCirc.pl

### [1.2] - 2017-10-12
### Fixed
- Improvement of using the standarded BED format with 12 fields for the reference gene annotation file

### Removed
- The "--sensitive" argument 


### [1.1] - 2017-06-30
### Added
- "-s/--seed" and "--sensitive" options to allow selecting different sizes of seed for initial mapping from unmapped reads.


### [1.0] - 2017-06-20
### Added
- Release v1.0


AutoCirc_v1.3.pl
#! /usr/bin/perl -w

###############################################################################
#Copyright 2017-2022 Chan Zhou
###############################################################################

#Program: AutoCirc is a computational tool to Automatically detect Circular RNAs (circRNAs) from RNA-Seq data. 
#website: https://github.com/chanzhou/AutoCirc 
#Author: Chan Zhou
#Contact: zhouchan99@gmail.com
#
#Reference:
#Identification and characterization of m6A circular RNA epitranscriptomes
#Chan Zhou, Benoit Molinie, Kaveh Daneshvar, Joshua V. Pondick, Jinkai Wang, 
#Nicholas O. Van Wittenberghe, Yi Xing, Cosmas C. Giallourakis, and Alan C. Mullen
#(submitted)
#BioRxiv version: http://biorxiv.org/content/early/2017/03/10/115899
#Last updated: 08/20/2018
#Pls read README before using this pipeline.

our $VERSION="version 1.3";

use strict;
use POSIX;
use warnings; 
use Getopt::Long qw(GetOptions);

my $USAGE =<<EOF;
Usage:
	[perl] $0 [options]

Options:
	
	-g/--genome     Absolute path of the reference genome file in fasta format (e.g. /genome_reference/hg19.fa) 
	-I/--gindex     Absolute path of the genome index of bowtie2 (e.g. /bowtie2_index/hg19 )
	--bam		Bam file of unmapped reads (output by bowtie/bowtie2/tophat)
	-b/--bed	Reference gene annotation in standard bed format 
			default: without reference gene annotation
			The absolute path to the annotated gene file
	-s/--seed	The size of the seed (e.g. anchor size)
			default: 20 (recommend values: 15-25; this value should be smaller than the (length of reads-10)/2)
	--mis		Maximum allowed mismatches during the search process (0, 1, or 2) 
			default: 0
	--min		Minimum distance between the two splice sites
			default: 100
	--max 		Maximum distance between the two splice sites
			default: 100000
	-o 		Output folder name 
			default: "AutoCirc_output"
	
	-h/--help	This help
	-v/--version	Print version

Examples :
	perl $0  -g /human_genome/hg19.fa -I /bowtie2_index/hg19 --bam unmapped.bam -b /hg19/Annotation/Refseq.gene.strandarded.bed --mis 0 --min 100 --max 100000 -s 20 -o autocirc_output 
	or
	perl $0  -g /human_genome/hg19.fa -I /bowtie2_index/hg19 --bam unmapped.bam -b /hg19/Annotation/Refseq.gene.strandarded.bed
	or
	perl $0  -g /human_genome/hg19.fa --bam unmapped.bam -s 15 -o autocirc_output (if do not have the reference gene annotations)   

NOTE: 
	1. The reference genome file should be the same one used in the previous Botwie/Bowtie2/Tophat mapping.
	2. The reference gene annotation should be in standard bed format with full 12 fields 
	   ( https://genome.ucsc.edu/FAQ/FAQformat.html#format1 )
EOF

sub main::HELP_MESSAGE {
	print $USAGE, $/;
	exit 1;
}

sub main::VERSION {
	print $VERSION,$/;
	exit 1;
}

my $unmapped_bam;
my $unmapped_sam;
my $genome_index;
my $ref_genome;
my $mismatch=0;
my $min=100;
my $max=100000;
my $ref_gene_bed="NA";
my $out_folder="AutoCirc_Output";
my $anchor_size=20;
my $help;
my $version;

GetOptions(
	    'bam=s' => \$unmapped_bam,
	    'gindex|I=s'=>\$genome_index,
	    'genome|g=s'=>\$ref_genome,	
	    'mis=s'=>\$mismatch,	
	    'min=s'=>\$min,
	    'max=s'=>\$max,
	    'bed|b=s'=>\$ref_gene_bed,
	    'seed|s=s'=>\$anchor_size,
	    'outfolder|o=s'=>\$out_folder,
	    'help|h' => \$help,
	    'version|v'=>\$version,
) or die $USAGE;

if ($help) {
	main::HELP_MESSAGE;
}

if ($version) {
	main::VERSION;
}

if ( (not $unmapped_bam) or (not $genome_index) or (not $ref_genome)) {
	
	main::HELP_MESSAGE;
}

if ($mismatch == 0 or $mismatch == 1 or $mismatch == 2){ }
else{
	main::HELP_MESSAGE;
}

if ((($ref_gene_bed ne "NA") and (not $ref_gene_bed =~/\S+\.bed/) ) or 
    ((not $ref_gene_bed =~/^\/\S+/) and (not $ref_gene_bed =~/^\~\/\S+/)) ){
	main::HELP_MESSAGE;
} 


if( (not $ref_genome =~/^\/\S+/) and (not $ref_genome =~/^\~\/\S+/)){
	warn "\nPls provide the absolute path of a reference genome file in fasta format\n\n";
	main::HELP_MESSAGE;
}
if( (not $genome_index =~/^\/\S+/) and (not $genome_index =~/^\~\/\S+/)){
	warn "\nPls provide the absolute path of the genome index of bowtie2\n\n";
	main::HELP_MESSAGE;
}

## step 1 take in bowtie/tophat unmapped bam/sam file
#         extract the anchors from two ends of unmapped-reads

if ($anchor_size>20 or $anchor_size<15){
	main::HELP_MESSAGE;
}

if($unmapped_bam =~/(\S+)\.bam/){
	$unmapped_sam=join "\.", ($1, "sam");
	warn "Now convert the BAM file to SAM file:\n";
	system "mkdir $out_folder";	
	warn "samtools view -h $unmapped_bam -o ./$out_folder/$unmapped_sam\n";
	GENE_SAM: my $generate_sam=`samtools view -h $unmapped_bam -o ./$out_folder/$unmapped_sam`;

	until($generate_sam ne 0){ 
		print "......... \n";
		sleep(60);
	}
}else{
	main::HELP_MESSAGE;
}

chdir($out_folder);
warn "cd $out_folder\n";


warn "Now generate the two end anchors of unmapped reads in fastq file\n";
warn "../script/unmapped2anchors.pl $unmapped_sam $anchor_size > unmapped_anchors.qfa\n";
my $generate_anchors=`../script/unmapped2anchors.pl $unmapped_sam $anchor_size > unmapped_anchors.qfa`;

until($generate_anchors ne 0){
	print ".........\n";
	sleep(20);
}

##step 2: bowtie2 map the anchores seq. into ref. genome

my $map_anchors;
$map_anchors=`bowtie2 -p4 --reorder --mm -D20 --score-min=C,-15,0 -q -x $genome_index -U unmapped_anchors.qfa 1>bowtie2.anchors.out.sam 2>bowtie2.anchors.log`;
warn "Now mapping the anchors into the reference genome\n";
warn "bowtie2 -p4 --reorder --mm -D20 --score-min=C,-15,0 -q -x $genome_index -U unmapped_anchors.qfa 1> bowtie2.anchors.out.sam 2> bowtie2.anchors.log\n";

until($map_anchors ne 0){
	print ".........\n";
	sleep(60);
}

##step 3  -- detect all circ meeting the GT/AG rule

warn "Now detecting the potential circRNAs following GT/AG rules in the back splice sites\n";
warn "../script/autocirc_gtag.v1.1 bowtie2.anchors.out.sam  $ref_genome $mismatch $min $max $anchor_size circ.gtag.bed linear.gtag.bed backsplice.raw.bed 2> clog.AutoCirc_gtag\n";
my $generate_gtagcirc=`../script/autocirc_gtag.v1.1 bowtie2.anchors.out.sam  $ref_genome $mismatch $min $max $anchor_size circ.gtag.bed linear.gtag.bed backsplice.raw.bed 2> clog.AutoCirc_gtag`;

until($generate_gtagcirc ne 0){
	print ".........\n";
	sleep(60);
}

##step 4 -- detect circ meet the annotated exon boundaries
if($ref_gene_bed=~/\S+\.bed/){
	warn "Now detecting the potential circRNAs meeting the exon boundaries of annotated genes\n";
	warn "intersectBed -a backsplice.raw.bed -b $ref_gene_bed -f 1 -wa -wb | cut -f 1-13,15,19-22 > backwards.annotatedexons.bed\n";
	
	my $generate_circ_annotatedexons=`intersectBed -a backsplice.raw.bed -b $ref_gene_bed -f 1 -wa -wb | cut -f 1-13,15,19-22 > backwards.annotatedexons.bed`;
	until($generate_circ_annotatedexons ne 0){
		print ".........\n";
		sleep(10);
	}

	warn "../script/filter_exon_boundariesv1.2.pl backwards.annotatedexons.bed 10 > circ.exonboundaries.unstrand.bed\n";
	my $filter_exonboundaries=`../script/filter_exon_boundariesv1.2.pl backwards.annotatedexons.bed 10 > circ.exonboundaries.unstrand.bed`;
	until($filter_exonboundaries ne 0){
		print ".........\n";
		sleep(10);
	}

	##step 5 -- combine the circ identifed based on GT/AG rule and these found by annotated exon boundaries
	warn "Now combining the circRNAs detected by GT/AG rules or exon boundaries of annotated genes\n";
	warn "../script/combine.pl circ.gtag.bed circ.exonboundaries.unstrand.bed > circ.combined.bed\n";
	my $combine=`../script/combine.pl circ.gtag.bed circ.exonboundaries.unstrand.bed > circ.combined.bed`;
	until($combine ne 0){
		print ".........\n";
		sleep(5);
	}
 

	##step 6: remove pseudo-circ

	warn "Now excluding the pseudo circRNAs which across two duplicated genes or multiple genes\n";
	my $rem_1=`intersectBed -a circ.combined.bed -b $ref_gene_bed -s -wa -wb | cut -f 1-12,14-16,18-21 > circ.ref_gene.bed`;
	until($rem_1 ne 0){
		print ".........\n";
		sleep(3);
	}

	my $rem_2=`../script/circ_across_mgenes.pl circ.ref_gene.bed > circ.pseudo.bed 2>err`;
	until($rem_2 ne 0){
		print ".........\n";
		sleep(3);
	}

	my $rem_3=`../script/my_join.pl -a circ.combined.bed -b circ.pseudo.bed -F 4 -f 4 | awk '\$9==\"\"' | cut -f 1-8 > circ.final.bed`;
	until($rem_3 ne 0){
		print ".........\n";
		sleep(3);
	}
	
	system "mkdir tmp";
	system "mv backwards.annotatedexons.bed ./tmp";
	system "mv circ.pseudo.bed ./tmp";
	system "mv circ.combined.bed ./tmp";
	system "mv circ.exonboundaries.unstrand.bed ./tmp";

}elsif($ref_gene_bed eq "NA"){ #do not input bed file 
	system "cp circ.gtag.bed circ.final.bed";
	warn "Without reference gene annotation, AutoCirc ab initio detect circRNAs based on GT/GT rules\n\n";
}
warn "Finished\n"
